src/presence/presenceModel.js

```diff
@@ -161,3 +198,70 @@ export const getPresenceStatusForUserId = (
     ? statusFromPresence(userPresence)
     : statusFromPresenceAndUserStatus(userPresence, userStatus);
 };
+
+//
+//
+// Reducer.
+//
+
+const initialState: PresenceState = {
+  byEmail: Immutable.Map(),
+};
+
+export function reducer(
+  state: PresenceState = initialState, // eslint-disable-line default-param-last
+  action: PerAccountApplicableAction,
+): PresenceState {
+  switch (action.type) {
+    case RESET_ACCOUNT_DATA:
+      return initialState;
+
+    case REGISTER_COMPLETE: {
+      // TODO(#5102): Delete fallback once we enforce any threshold for
+      //   ancient servers we refuse to connect to. It was added in
+      //   #2878 (2018-11-16), but it wasn't clear even then, it seems,
+      //   whether any servers actually omit the data. The API doc
+      //   doesn't mention any servers that omit it, and our Flow types
+      //   mark it required.
+      const data = action.data.presences ?? {};
+      return {
+        byEmail: Immutable.Map(data),
```

Comment:
Hmm, should this style-guide entry—

> **Don't construct an `Immutable.Map` with an object-as-map**: Whenever you
> create a non-empty `Immutable.Map`, pass an array of entries instead of an
> object-as-map. For example:
> ```js
> Immutable.Map([[HOME_NARROW_STR, [1, 2]]]) // good
> 
> Immutable.Map({ [HOME_NARROW_STR]: [1, 2] }) // BAD -- don't do
> ```
> 
> This is essential in order to get effective type-checking of the value
> passed to the `Immutable.Map` function when a key is computed (like
> `HOME_NARROW_STR`) instead of literal (like 'foo'). In the Immutable type
> definition, the object-as-map branch involves an object type with an indexer
> property, and in general [Flow seems to have trouble with those][].
> 
> [Flow seems to have trouble with those]: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.60Immutable.2EMap.28.7B.20.5BSTRING_CONST.5D.3A.20.E2.80.A6.20.7D.29.60.20busted.20by.20Flow.20bug/near/1481088

—only forbid passing a _literal_ object-as-map? If I assert that the object-as-map `data` has `number` values instead of `UserPresence` values, Flow does complain about that.

Oddly I don't get a Flow error with this:

```diff
diff --git src/presence/presenceModel.js src/presence/presenceModel.js
index a46559acb..9a3e8f2e3 100644
--- src/presence/presenceModel.js
+++ src/presence/presenceModel.js
@@ -225,7 +225,7 @@ export function reducer(
       //   mark it required.
       const data = action.data.presences ?? {};
       return {
-        byEmail: Immutable.Map(data),
+        byEmail: Immutable.Map<string, number>(data),
       };
     }
```